feat: Non MFS Files API tutorial

dominguesgm opened the pull request:
Introducing a non-MFS Files API Tutorial. Currently consisting of 6 lessons, 1 simple and 5 code lessons.

Structure:
1. Introducing the File API
2. Adding a file
3. Read the contents of a file
4. Add files in a folder
5. Listing the files in a directory
6. Getting all the files in a directory tree

Contents should still be proof-read and validated before merging.

Closes #203

dominguesgm commented:
Review made by @ericronne is in #304. I'll be checking it out today

terichadbourne commented:
This is looking good @dominguesgm! 

Some outstanding items include: 
[ ] update the featured tutorials so this shows on the homepage
[ ] add the resources page contents

Before I do any text edits, I want to ask some bigger conceptual questions and share areas where I got confused on my first pass through this so you can update the content accordingly. Note that I have only run through this as an end user and have not yet looked at your validation code. 

**Lesson 1:** 

I suspect this section may need a little tweaking if my understanding is correct: 
```
If you've gone through our Decentralized Data Structures tutorial — or even Blogging on the Decentralized Web — you already know you can store primitives, objects and arrays in the network.

Storing these types of data is an interesting, but limited, use case. What if you want to share a picture of a kitten? How would you upload it to the network and provide a way for your friends to see it? What about a larger file, such as a funny video? How should the file be placed in the Directed Acyclic Graph (DAG) — in a single block or split into chunks?

You can think of the File API as an abstraction layer above the DAG API. The File API prepares files to be placed in the network, and ensures that IPFS knows how to access them. The details of what this API actually does will be covered later in this tutorial.
```
My understanding has been that DAG isn't a _more limited_ use case, it's a _broader use case_ because it can deal with all different kinds of data (including the obscure ones you mention), but that it's just really obnoxious to use for the most common use cases, like just sharing files. You _could_ use the DAG API to share files, but you'd have to do a ton of extra work yourself to manipulate the file objects in the way you'd need to. The File API uses the DAG API under the hood but has lots of extra functionality built in to make it work with files more conveniently. We could have done this on our own, but the File API saves us time. (@mikeal can you please confirm this paragraph is true? ) 

**Lesson 2:** 
- **Working with files in ProtoSchool:** In shortening the information about how file upload works in ProtoSchool, you've left out the very important detail that we make the files you upload available to you in the `files` array in the code editor, and you're not introducing what information the browser has available to it (file.name, etc.). This gets called out in passing in lesson 4 but I think we need that referenced earlier, before anyone sees `const run = async (files) => {` at the top of every lesson. I feel like it might actually be worth repeating lesson 3 of MFS here (which displays to the user what details are made available), or at least pulling more information from it than you have, but please let me know what you think. If we repeated it then it should be a different lesson from the one teaching `add` and the lesson numbers would change.
- **The `add` method:** In your "Inspect results" section, I'd love clarification that this is the value returned by the `add` command, and a reminder that the object returned by IPFS for each file contains a "hash" property that contains its CID, which will come in handy in our next lesson. (I think the hash is the CID based on your note earlier in the lesson but I see that path and hash are identical here so it could be path as well. We should explain what makes them different. ¯\_(ツ)_/¯ 
- In both lessons 2 and 4 you mention that you can pass either a single file object or an array of file objects. Let's include examples of the two formats with values filled in (`kitten.jpg`, not `file.name`). Documentation sometimes suffers from not giving clear examples of how to use things, so including a humanized example of "if I had a photo saved as 'kitty.jpg', I could add it like this:" is very helpful. And then it could be done again with an example of an array of dog and cat photos, etc. You can find some examples of this format here: http://localhost:3000/#/mutable-file-system/04

**Lesson 3:** 
- As noted above we need a reminder of how we find out the CID of the file, not just how we use it, so you could reference the results from lesson 2 near the start of this one. Something like "As we saw in the previous lesson, the `add` method returns an object for each file that includes a `hash` property containing it's CID, blah blah"
- My first attempt at this lesson made the validation hang because I passed in the CID without putting quotation marks around it, so it would have tried to read it as a variable and not a string. I'd recommend both adding a hint about this and seeing if there's any way for you to catch that specific error in validation. The documentation technique described above would also help, so I had an example where I saw an actual CID in quote marks. 
- In general I think the concept of overwriting values is harder than the concept of creating a new variable to hold something you've messed around with, so I'd avoid something like this: 
```javascript
let message = await ipfs.cat('QmWCscor6qWPdx53zEQmZvQvuWQYxx1ARRCXwYVE4s9wzJ')
message = message.toString('utf8')
```
- What are your thoughts on whether we should remove the file upload component here, since it's not actually being touched? Normally it's handy to have the files still in your browser even when they're not needed for this one, but I think it adds unnecessary complexity to the code challenge. I also had to redo my file uploads for lesson 4 anyway because i wanted to see what it was like to add multiple files to a directory and i'd only done one previously. Let me know what you think. 

**Lesson 4:** 
- Feedback from previous reviewers of other tutorials was that we should consistently say directory rather than folder. 
- The first time I read the documentation for the add `method` I took `path` to mean a path on my computer where I was getting the file from, not a destination path I was inventing, so that may be worth calling out. 
- Conceptually I think we need to spend some time explaining what directories are in IPFS. I believe that a directory's hash is going to change each time we change its contents, which is important to highlight. Also, how are we able to name directories? Is this some kind of weird abstraction like MFS is? We just told everyone that the regular Files API can't act like a file system, so what's the catch? Is it that we'll never be able to find it again by its name and will always have to find it by its CID, which keeps changing? (We may need to call in an expert to help with this question, maybe you can chat quickly with Hugo?)
- You say "Don't forget, you can get the name of the file you uploaded with file.name." but you've never told them that's possible, as noted earlier. 
- Since not everyone who uses the site is a JavaScript pro, they may be more familiar with `forEach` than with Array methods like `map` or `filter`. In some of the other tutorials we've provided solutions that showed both a more advanced and an easier option, with one of them commented out. As an example, this is what I tried on my own without looking at your solution: 
```javascript
/* global ipfs */
const run = async (files) => {

  let fileObjects = []
  files.forEach(function (file) {
    let fileObject = {
      path: "/dir/" + file.name,
      content: file
    }
    fileObjects.push(fileObject)
  })

  let result = await ipfs.add(fileObjects, {wrapWithDirectory: true})

  return result
}
return run
```
- I think that what you're listing under "Inspect results" is the value returned by the `add` function, which should be phrased a bit more clearly (as opposed to this being you showing us an `ls`, for example. When I did this I added two files and got 4 objects in the returned array, which needs some further explanation, I think probably two were for the files I added, one was for the path "dir" so presumably the directory and one was for path "" so maybe the root node? ¯\_(ツ)_/¯ 


**Lesson 5:** 
- I don't think it's clear what you're doing to find the pathCID, in part because it's a prerequisite here that I be familiar with the `find` method, which isn't anything I use regularly. Are you searching through the results from `add` to find one where the path is the name of the directory, and then getting the hash value from that same object? Do you want to explain this or will users need to learn to do it enough that it should be its own lesson, or part of the challenge in this one? Again, we probably need some context to connect the results of the last lesson to the start of this one.
- Your solution code doesn't pass this lesson, which would make it fail Cypress testing if Cypress knew how to test file upload lessons. :) Looks like you forgot to include `pathCID` and just used `await ipfs.ls()`. The error message from your validation is: "The CID provided to ipfs.ls is incorrect. Make sure you're using the pathCID variable we provided."  I assume that doing  `await ipfs.ls()` will either flat out fail or show the root directory contents, either of which is probably worth calling out as a separate error from using a CID that's wrong in some other way. 
- Again I'd specify how you're getting the "inspect results" display, so perhaps changing "Here are the contents of the dir directory." to "Here are the contents of the dir directory as returned by the ls method."

**Lesson 6:** 
- I find it pretty hard to understand what you're asking the learner to return in this coding challenge, which combines a couple of different concepts.  
- I'm also struggling to come up with the right method to complete this task without using `map`, which I consider more advanced. For previous lessons we've added links to JavaScript documentation when we're inadvertently testing JavaScript skills instead of IPFS skills.
- Again, I don't think we should expect users to overwrite variable names. 
- If we're going to use this format then I think we need to cut out the file upload here because the CID they're getting from and the files being returned have nothing to do with those they've uploaded. Another alternative would be to have them add one new file from a CID (could be the success.txt one) to the directory they already created. This would have a bonus (I think) of showing how the CID of the directory changes when its contents change. Obviously we can't include the `toString('utf-8')` conversion if we do that, since we wouldn't know what kind of files they had. Tell me more about why you asked them to do that.... because it's a skill they'll need (although the did it earlier in another context) or because displaying the results was unwieldy when the content filed wasn't a string? 
- Again we could make the phrasing under "Inspect results" a little more specific.

In going back to the [original issue](https://github.com/ProtoSchool/protoschool.github.io/issues/203) and outline for this tutorial, the only suggestion I don't see covered here is to show people how to use the CID to view a file on the gateway. I believe neither the files they're adding themselves nor the ones you've used as examples are living there, so maybe that's not practical anymore? Thoughts? 

I remember hearing about some MFS pitfalls in [the end section of Alan's course at IPFS camp](https://youtu.be/Z5zNPwMDYGg?t=3134) that highlighted important distinctions between MFS and all the rest. Would you mind taking a quick look to see whether there are any bits here worth including when you describe the difference between the regular File API and MFS? 

Sorry for the novel here, @dominguesgm. I'll be out on Monday so just want to share as much feedback as I can so you can get started on anything that doesn't require more clarification. Happy to schedule some time to chat through any of this feedback if you disagree or if I haven't been clear. 

Thank you so much for building this tutorial. It's fabulous!

dominguesgm commented:
Thank you for the in depth feedback @terichadbourne , I have some points I'd love to clarify/discuss, which I will structure by lesson:

** Lesson 1: ** 

* I agree with your point, I'll probably rephrase the way I'm going about comparing the DAG API and the Regular File API. I was under the impression you didn't think we should go into more complex details and that's why I avoided discussing things the Regular File API introduces, like file chunking, etc. (if I'm not mistaken). Maybe it would be relevant to mention these advantages briefly

** Lesson 2: **

* I've removed the section on file upload and copied the lesson 3 from the MFS tutorial to this tutorial (right before this lesson). I'll probably rework its content because I don't recall it talking about the file properties in the lesson itself.

* I've changed the success message to tell the user the output in the `Inspect Results` corresponds to the value returned by the `add` command.

* I've also added an example with a friendlier variable name and a more concrete scenario as suggested, right after presenting the `add` method

** Lesson 3: **

* We could do this, but I think we should discuss the pro's and con's: simplify interface and the code submission vs improving the continuity between lessons. #242 presents a similar suggestion for a lesson in the MFS tutorial

** Lesson 4: **

* I replaced all instances of folder with directory, but missed that one 😅 

* I'll try to make the meaning of path clearer, and I agree we should talk to someone with a better understanding of the inner workings of the API about what directories actually are.

* The `file.name` property will be mentioned in the new Lesson 2

* I can change the `map`s usage and try to find a simpler way to approach this. I'll probably go for a simple `for` loop, I feel like `forEach` would be at practically the same level as `map` as they're both array functions, but I might be wrong. I'll take a look at the solutions you mentioned which showcase an advanced and an easier approach.

* As in the other lesson, I'll work on a better success message, though maybe I'll write the explanation for the extra elements in the array resulting from the `add` method right before the code exercise.

** Lesson 5: ** 

* My approach for this code lesson was to write the code needed to setup the exercise as concisely as possible in order to avoid having a very large boilerplate for the exercise. Maybe I can put a comment explaining what that block of code is doing? Basically I'm adding all the files into a `dir` directory and then `find`ing the `CID` of the directory so the user can then `ls` it afterwards.

* Yes, the issue was the `ls` method was being called with no argument. It will be fixed in the next commit.

* I agree with this last point on a clearer success message, again. I'll go through the success messages of the coding exercises.

** Lesson 6: **

* I understand your point, and I did spend some time debating on whether the exercise was becoming too complex or not.

* About the overwriting of the variables, if I change to that approach, I would probably remove the return statement from the code boilerplate and replace it with a comment saying something along the lines of "don't forget to return the result". This because if we don't expect the variables to be overwritten, we can't know for sure which variable will contain the results the user has to return. Another option would be to have something like `return // don't forget to place your result here`, but I really dislike the idea of leaving an empty return in, it makes it seem like that part is already done.

* This exercise is more complex because my approach was to teach how `get` would return the whole directory structure a `CID` points to. The problem with using user files here is that we don't control which files the user is uploading. For example, I always tested the exercises with `jpg`s ranging from 100kb to 500kb. The get command returns, among other data, a buffer of the file's contents. This, and because I wanted the user to be able to read the contents of the files in the directories, is why I pre-built a directory structure in the node for this exercise. I think we could remove the upload file functionality from this exercise, as well as the boilerplate code which is not used, to avoid confusing the user.

I'll be honest, the CID explorer went over my head, I can try to look into it. But from what I understand, we may have some issues letting a user upload a file and then asking them to view it on the CID explorer because the page may hang due to not finding the file at all. I'll still look into it.

Sorry for the long message, but hopefully this will help us get closer to a final version of the tutorial 😃

terichadbourne commented:
@dominguesgm Thanks for working on some edits! A few responses to some of your points below. Let me know if I've missed anything you want to be sure to discuss. 

**Lesson 1:** 

> I agree with your point, I'll probably rephrase the way I'm going about comparing the DAG API and the Regular File API. I was under the impression you didn't think we should go into more complex details and that's why I avoided discussing things the Regular File API introduces, like file chunking, etc. (if I'm not mistaken). Maybe it would be relevant to mention these advantages briefly

Yeah, apologies if I led you astray with our earlier convo on this. I think it's okay to present some bonus advantages that the File API offers (briefly in passing) even if they won't be covered explicitly in this lesson, if they help to explain why you would choose this over the DAG API. For example, (if it's true 😂) you could mention that the Files API handles the complexity of splitting your file into chunks of appropriate size and provide a link to whatever the best documentation is that we know of about how that works. And then if, for example, there's an option you can send to the `add` method that tells it to use something other than the default chunk size, you could mention that in passing in the `add` lesson with a link to the documentation that shows you how to do it. (Note that I'm just inventing hypothetical examples here; I don't know how chunking actually works.)

**Lesson 2:**

> I've removed the section on file upload and copied the lesson 3 from the MFS tutorial to this tutorial (right before this lesson). I'll probably rework its content because I don't recall it talking about the file properties in the lesson itself.

MFS lesson 3 talks about the `files` array in the content of the lesson and then talks about the properties that browser file objects contain in the part that appears after you submit a correct result, so that you see an example from the file you just uploaded. Feel free to propose another approach if you prefer. 

**Lesson 3:**

> We could do this, but I think we should discuss the pro's and con's: simplify interface and the code submission vs improving the continuity between lessons. #242 presents a similar suggestion for a lesson in the MFS tutorial
Yes, it's definitely a toss-up. I think one main difference between this and the other example is that we ended up integrating the files we put in with the files the user put in over the course of the MFS lesson and we don't really do it here? We can chat more about this one live. 

**Lesson 4:**
> I'll try to make the meaning of path clearer, and I agree we should talk to someone with a better understanding of the inner workings of the API about what directories actually are.

👍 I have a couple of folks in mind who we can ping for a review and improved explanations after your current batch of revisions. 

> I can change the `map`s usage and try to find a simpler way to approach this. I'll probably go for a simple `for` loop, I feel like `forEach` would be at practically the same level as `map` as they're both array functions, but I might be wrong. I'll take a look at the solutions you mentioned which showcase an advanced and an easier approach.

I personally am more comfortable with `forEach` than with any of the other, more specifically-purposed array methods, but I'm also happy to see a simpler solution with `for`. Let me know what you think works. We got another issue filed today where array methods were the blocker for someone who's not used to working in JavaScript, and where we planned out more potential variable names than everyone would want to use. We're always going to have the battle where folks who are strong in JavaScript will think we're being ridiculously verbose if we avoid those methods and beginners won't understand us if we use them, so I do think offering alternatives in the solution is reasonable for now. 

It would be interesting to offer a toggle feature in the future where you could view solutions in "advanced" or "beginner" JavaScript mode. 

**Lesson 5:** 

> My approach for this code lesson was to write the code needed to setup the exercise as concisely as possible in order to avoid having a very large boilerplate for the exercise. Maybe I can put a comment explaining what that block of code is doing? Basically I'm adding all the files into a `dir` directory and then `find`ing the `CID` of the directory so the user can then `ls` it afterwards.
I do think that whether or not we make them practice it themselves, we need explicitly teach users how to figure out the CID of their directory, assuming that's something someone would do in real life without MFS. It could be narrative in the lesson and then a comment next to the more advanced JS in your code, potentially, if we're not surfacing it as a separate exercise, and I think it needs a callout when the results of the previous lesson are shown, as noted earlier. 

**Lesson 6:**


> About the overwriting of the variables, if I change to that approach, I would probably remove the return statement from the code boilerplate and replace it with a comment saying something along the lines of "don't forget to return the result". This because if we don't expect the variables to be overwritten, we can't know for sure which variable will contain the results the user has to return. Another option would be to have something like `return // don't forget to place your result here`, but I really dislike the idea of leaving an empty return in, it makes it seem like that part is already done.

Honestly, we're never going to know what people will want to do with their variables.  My guess is that folks who are super new to JavaScript will most appreciate them being provided and well named, in which case they'd want the multiple variable names and distinct steps (without fancy overwriting array methods). Folks who are advanced will be able to ignore us as much as they want to and remove unnecessary steps, so if we had to favor one side of this argument I'd favor the total newbie. However, it might also be okay not to provide any blank variable statements, so people can use as many or as few steps as they want, and I don't mind `return // don't forget to place your result here` or leaving off the return and including a "don't forget to return your result warning." As long as we can catch "You forgot to return a result" in the validation, I'm flexible.

> This exercise is more complex because my approach was to teach how `get` would return the whole directory structure a `CID` points to. The problem with using user files here is that we don't control which files the user is uploading. For example, I always tested the exercises with `jpg`s ranging from 100kb to 500kb. The get command returns, among other data, a buffer of the file's contents. This, and because I wanted the user to be able to read the contents of the files in the directories, is why I pre-built a directory structure in the node for this exercise. I think we could remove the upload file functionality from this exercise, as well as the boilerplate code which is not used, to avoid confusing the user.

If I'm understanding you correctly, it sounds like using `get` on an unknown file type (whatever the user uploads) without overwriting the results with non-Buffer will cause something totally unreadable _and_ we won't have any known quantities to do validation against, and that's why you provided your own files. How bad would it be visually if you showed the buffer results of the .txt files without stringifying them first? 

Option A (if it doesn't print a giant mess to the screen): 
* Just `get` the directory contents (provided by us) in one lesson, and see how useless the returned buffers are
* In the following lesson, reveal the contents of a specific text file 

Option B
* Get a specific file (not a directory) and reveal its content (I think this is what was proposed in the original issue)

The thing I'm trying to avoid is the JavaScript trick of overwriting all the buffers with strings. I think I would find it an easier JavaScript challenge, for example, to find the content of the text file titled X. But perhaps if you give an example that uses simpler JavaScript methods I'll change my mind. :) 

Agreed that if we'll only use the files you provided we should skip the file upload. 

> I'll be honest, the CID explorer went over my head, I can try to look into it. But from what I understand, we may have some issues letting a user upload a file and then asking them to view it on the CID explorer because the page may hang due to not finding the file at all. I'll still look into it.

I wasn't meaning to reference the CID explorer tool specifically, just the general concept that you can find things by CID on the gateway. I believe you just need to go to https://gateway.ipfs.io/ipfs/<your_CID_here>, but I think it does depend on the gateway being fully functional and enough people having pinned a thing. Let's ignore this suggestion for now.

dominguesgm commented:
@terichadbourne I made a commit with changes covering most of the topics we discussed. There are still a few I'd like to discuss more in depth:

* a better description of what directories are
* a way to simplify the last lesson

terichadbourne commented:
I opened an issue for the toggle idea, although that is definitely not a priority at the moment. Feel free to add any ideas there: #312 

Look forward to chatting later today about those remaining issues.

terichadbourne commented:
@alanshaw A few specific content questions for you as @dominguesgm puts the final touches on this... 

If one uses `add` and passes in an array of files with `{ wrapWithDirectory: true }`, should they be able to later add more things to the same directory and have them all end up together, or is its existence basically all in our heads? We want want to make sure we're correctly describing what this concept of a directory really is within the non-MFS system. Any tips on how best to explain it in [lesson 5](https://github.com/ProtoSchool/protoschool.github.io/pull/303/files?short_path=101675f#diff-101675fe70062bedfee724d302f2571a), and perhaps contrast it to what happens in MFS for those who already did that tutorial? (Also, if you were going to be able to add more files to a directory, I presume we'd want to point out that its CID changes when you do that.)

We could also use a closer look at [lesson 1](https://github.com/ProtoSchool/protoschool.github.io/pull/303/files?short_path=c2f6ec5#diff-c2f6ec5f1a7623e875fb05ef3ad906c0) where Gil explains both the difference between the Files API and the DAG API and between the Files API and MFS. Want to make sure we're highlighting the right pros/cons/functionality here without getting too deep. 

Feedback on all aspects of the tutorial are welcome, those are just our biggest newbie questions at the moment. :)

dominguesgm commented:
I wanted to put some more context into the question about adding multiple files into a directory with the `add` method.

I was experimenting with it and tried to add two different files into the same directory, with different file names, in **separate** `add` calls. However, if I called `get` with the CID to the common directory, it would only retrieve the directory and the last added file.

This is the code I was trying to run:

```javascript
const IPFS = require('ipfs');

const ipfs = await IPFS.create();

let res = await ipfs.add({ content: Buffer.from('HelloWorld'), path: '/dir/file.txt'}, {wrapWithDirectory: true});

res = await ipfs.add({ content: Buffer.from('hello world 2'), path: '/dir/file2.txt'}, {wrapWithDirectory: true});

console.log(await ipfs.get(res[2].hash));
```


Which returns an array with the root, `dir` directory, and last added file:

```javascript
[ { hash: 'QmPrjMFMsjwpgn5EVwY7k9Qrm5JJT47Q29TW5XxjTgtMZ9',
    path: 'QmPrjMFMsjwpgn5EVwY7k9Qrm5JJT47Q29TW5XxjTgtMZ9',
    name: 'QmPrjMFMsjwpgn5EVwY7k9Qrm5JJT47Q29TW5XxjTgtMZ9',
    depth: 1,
    size: 0,
    type: 'dir' },
  { hash: 'QmfXcmsEgF2PeevmbaoS1jGuNMBhYRfm468LW59qGdQGoA',
    path: 'QmPrjMFMsjwpgn5EVwY7k9Qrm5JJT47Q29TW5XxjTgtMZ9/dir',
    name: 'dir',
    depth: 2,
    size: 0,
    type: 'dir' },
  { hash: 'QmUGTcXwYTqHd5hbkRELDf6uSnBRjYv8CDEjsNDYde6o8h',
    path:
     'QmPrjMFMsjwpgn5EVwY7k9Qrm5JJT47Q29TW5XxjTgtMZ9/dir/file2.txt',
    name: 'file2.txt',
    depth: 3,
    size: 13,
    type: 'file',
    content: <Buffer 68 65 6c 6c 6f 20 77 6f 72 6c 64 20 32> } ]
```

alanshaw commented:
> ```js
> const IPFS = require('ipfs');
> 
> const ipfs = await IPFS.create();
> 
> let res = await ipfs.add({ content: Buffer.from('HelloWorld'), path: '/dir/file.txt'}, {wrapWithDirectory: true});
> 
> res = await ipfs.add({ content: Buffer.from('hello world 2'), path: '/dir/file2.txt'}, {wrapWithDirectory: true});
> 
> console.log(await ipfs.get(res[2].hash));
> ```
> 
> Which returns an array with the root, `dir` directory, and last added file:

Exactly, you're essentially double-wrapping with this code: `/wrapper/dir/*`, but it allows you to access the content via `/ipfs/QmWrappingDirectory/dir/file.txt` if you so wish.

dominguesgm commented:
Thank you @alanshaw for such an in depth explanation of what functionality the wrapWithDirectory option is supposed to provide, I was not interpreting it correctly!

terichadbourne commented on src/tutorials/0005/02.md:
```diff
@@ -0,0 +1,12 @@
+As a matter of security, web browsers don't let us directly change files that live in your computer's file system. Therefore you'll need to upload one or more files to the browser that you can use throughout this tutorial.
```
I addressed this in my latest edits. Thanks!

terichadbourne commented:
@dominguesgm I've reviewed lessons 1-3 and the resources page and pushed commits that include:
- copy edits to Lessons 1-3 (lessons, exercises, and additional text that pops up with results)
- changing "File API" to "Files API" to match the documentation
- adding to the resources

Let me know if there are any edits you disagree with or would like to discuss further.

I didn't make it to lesson 4, but am happy to proof that one on Monday if you still think it'll be stable. I know you're still making substantive edits to the following lessons, so I'm intentionally not reviewing them in detail yet. 

Note that I have not yet looked closely at your code validation, just the copy.

terichadbourne commented:
@alanshaw & @mikeal We'd love your input on the description [here](https://github.com/ProtoSchool/protoschool.github.io/pull/303/files?short_path=c2f6ec5#diff-c2f6ec5f1a7623e875fb05ef3ad906c0 ) of how the non-MFS Files API differs from the DAG API and from MFS. Trying to get at when each one would be most useful and address the most important advantages / disadvantages at a high level without getting into the weeds on anything.

We can expand on a few more specific points (like how `{wrapWithDirectory: true}` leads to something quite different from MFS) in later lessons where it seems relevant enough to explain to beginners.

We'd also love suggestions on other good resources (beyond the docs) that we can suggest for learning about the non-MFS Files API beyond the docs. 

And @alanshaw what are we supposed to be calling this non-MFS thing? Currently it's referenced as the Regular Files API when MFS is mentioned in the same breath or the Files API when it's not, but I know it doesn't use the `files.someMethod` prefix. 🤷‍♀️ (@dominguesgm this lack of `files` preceding the method might actually be worth pointing out for folks who've already done the MFS tutorial.)

terichadbourne commented:
@dominguesgm I've just pushed a commit with some text edits to lesson 4. 

Note that I'm making sure never to use the word "upload" when discussing IPFS. That verb implies that I'm taking a file that was on my computer and putting it somewhere in the cloud, whereas what we're actually doing with IPFS is leaving the file on our computer but more or less applying a label to it to make it part of the network. So I'm instead choosing verbs like "add" or "store."

The reason I removed your code block with the output of the `add` command is that it was covered in both the body and the results of the previous lesson. Referencing the relevant fields is definitely important, but I tried to condense the mention a bit. 

I believe there's still some validation work to be done on this one. There's no `else` clause, so the submit button is just spinning with any wrong answer I try, for example if I: 

return the wrong string somehow
```javascript
return "some string" // having deleted all the other code
``` 
or if I fail to include the quote marks around the CID
```javascript 
return (await ipfs.cat(QmWCscor6qWPdx53zEQmZvQvuWQYxx1ARRCXwYVE4s9wzJ)).toString('utf8')
```
of if I use the write code but try a CID of my own generated in the previous lesson (whether or not the file is a `.txt` file) 

or if I use correct code but remove one `{` or `(`, a syntax error that would normally get surfaced if none of our own errors are caught

or if try to get fancy and smush the two methods together without wrapping the await statement
```javascript
return await ipfs.cat('QmWCscor6qWPdx53zEQmZvQvuWQYxx1ARRCXwYVE4s9wzJ').toString('utf8')
```
(I added a second example to the lesson to show the necessary wrapping, since that's something I've screwed up in past lessons.)

I suspect some of these might be able to have their own failure messages in `else if` clauses, and some should be caught by a final `else` clause with a more generic failure message. 

I also edited the tutorial lists so we only feature 4 options. 

Thanks for your continued work on this!

terichadbourne commented:
I've pushed some copy edits to Lesson 5 and have some additional feedback/questions @dominguesgm: 
- Great job making it clear that these directories aren't the thing we normally think of as directories!
- Your attempt to be simpler than `Array.map` here is to use iterators, which weren't introduced until ES6. I suspect `forEach` would be more familiar to JavaScript novices, but I could be wrong. We should be providing hints linking to the JS documentation for the mostly likely couple of methods to be used.
- I haven't looked super closely at the validation, but I noticed a reference to `dir` which is probably left over from an old version. 
- Also in the validation, you might want to check for multiple slashes in the filename in case someone like me gets curious and tries to use a subdirectory instead of the root, saying something like "it looks like you added those files to a subdirectory; can you add them to the root directory instead?" Or maybe we just need to change the existing message, if you think that's the only error that the clause might catch.
- The note in the results about me needing to save the CIDs makes me think I'll need them in a future lesson. If that's not the case, we might want to adjust the wording.
- It looks to me like the validation again has `if`s and `else if`s but doesn't have a catch-all `else` to return a generic message.

Let me know if you have questions or concerns about any of my edits. Happy to discuss!

terichadbourne commented:
@dominguesgm I've added some lesson 6 copy edits for you, including a look at the format of data returned by `ls` so you can call out in lesson 7 that the format of `get` results is identical to that of `ls` except for the `content` field added w/ files. Let me know if you have questions or concerns.

terichadbourne commented:
@dominguesgm I've just pushed some minor copy edits to Lesson 7, including reducing the number of files passed into your sample directory just to cut down on the size of this lesson. (I'm assuming these are fake CIDs so it doesn't matter that I changed the contents?)

I also updated the output from lesson 6 to create a better lead-in to this content.

Can you check to see if the value return matches the buffered file contents instead of just assuming anything that's not a string is the buffered file contents? I hit that same clause by doing this, so we're probably missing some conditionals: 
![image](https://user-images.githubusercontent.com/19171465/67581782-9dc84600-f716-11e9-9728-4d44c7ed6243.png)


I think we should plan ahead and create custom error messages for these two specific mistakes: 

Using only the CID of the directory without adding the filename to it:

![image](https://user-images.githubusercontent.com/19171465/67581229-ab310080-f715-11e9-81cc-c252c09ba0b3.png)

Using the filename without putting the directory's CID before it: 
![image](https://user-images.githubusercontent.com/19171465/67581380-e7fcf780-f715-11e9-9431-e99f7c14f340.png)

I'm not sure if you'll cover it in Lesson 8, but if not, it might be useful to at some point show that if you have nested directories, you'll have more than two options for retrieving a file.

terichadbourne commented:
@dominguesgm I've added some lesson 8 copy edits, the most notable of which is adding in a comparison between `get` and `ls`, not just `get` and `cat`.  I also reviewed your lesson 7 edits and made sure the third naming option was covered in the body of the lesson, not just the code challenge.

I also updated a few things in Lesson 1 based on the response we got from @alanshaw about how we should refer to the non-MFS Files API (as the Regular Files API) and removed the part about the Files API being an abstraction on top of the DAG API. 

Let me know if you have questions or concerns re any of my edits. Happy to discuss!

dominguesgm commented on src/tutorials/0005/04.md:
```diff
@@ -22,16 +22,4 @@ If we knew the path to a file in IPFS we wanted to read, we could call the `cat`
-
-The IPFS path can take several formats (you can read about them in the [File API](https://github.com/ipfs/interface-js-ipfs-core/blob/master/SPEC/FILES.md#cat) documentation). In this lesson, we will use the simplest one: the `hash` string present in the objects we receive when the `ipfs.add` function is called.
-
-The returned data comes in the form of an `Buffer`. If you uploaded a plain text file (such as a `.txt`), you can read the contents of the file by calling `result.toString('utf8')`.
```
Isn't it important to explain the `toString` method in the lesson's body?

terichadbourne replied:
It's still there. I was removing duplicate contents after some sort of failed merge situation. :) Just further fixed my original edits in a new commit to replace a missing single-line example.

terichadbourne commented on src/tutorials/0005/01.md:
```diff
@@ -19,7 +19,7 @@ The Files API, on the other hand, is custom-built for a more specific use case.
 If you've read our [Mutable File System tutorial](https://proto.school/#/mutable-file-system), you may be thinking, "I've already learned how to work with files on IPFS. How will this be any different?"
 
-The Mutable File System (MFS) provides an API designed to replicate familiar file system operations such as `mkdir`, `ls`, `cp`, and others, mimicking the way you organize files and directories on a computer. However, the way that content is addressed in IPFS makes it an immutable file system. The address to a file or directory depends on its contents, so any change to a file or directory will re ...
+The Mutable File System (MFS) provides an API designed to replicate familiar file system operations such as `mkdir`, `ls`, `cp`, and others, mimicking the way you organize files and directories on a computer. However, the way that content is addressed in IPFS makes it an immutable file system. The address to a file or directory depends on its contents, so any change to a file or directory will re ...
```
Nice catch @dominguesgm. This was actually my bad; I meant to remove that whole sentence because I shared this info further down. Just included the fix in another commit.

terichadbourne commented:
@alanshaw This PR is now ready for a full review, and includes edits based on our chat. Would it be possible for you to take a look and provide your feedback between now and Thursday?

alanshaw commented on src/tutorials/0005/07.md:
```diff
@@ -0,0 +1,68 @@
+So far, we've used the CID of the file or directory as the path when accessing a file. However, now that we've learned we can wrap a number of files into a directory, we have a new way to address a file.
+
+As you saw in our lesson on the [wrapWithDirectory option](https://proto.school/#/file-api/05), we can add one or more files to a new directory and, when doing so, we need to provide a name to each one of the files we add. As a result, the `add` method call returns us an array which contains the `hash` value (CID) for each of the files and directories created.
```
```suggestion
As you saw in our lesson on the [`wrapWithDirectory` option](https://proto.school/#/file-api/05), we can add one or more files to a new directory and, when doing so, we need to provide a name to each one of the files we add. As a result, the `add` method call returns us an array which contains the `hash` value (CID) for each of the files and directories created.
```

alanshaw commented on src/tutorials/0005/07.md:
````diff
@@ -0,0 +1,68 @@
+* Using the `hash` (CID) of the file's containing directory (`lists`), followed by the file's path relative to that directory:
+
+```javascript
+await ipfs.cat("QmPT14mWCteuybfrfvqas2L2oin1Y2NCbwzTh9cc33GF1t/kitty-pic-list.txt")
````
Really this should be prefixed with `/ipfs/`

terichadbourne replied:
Do you mean it's supposed to be `await ipfs.cat("/ipfs/QmPT14mWCteuybfrfvqas2L2oin1Y2NCbwzTh9cc33GF1t/kitty-pic-list.txt"`)?  Tell me more! 

This is what I see in [the `cat` documentation](https://github.com/ipfs/interface-js-ipfs-core/blob/master/SPEC/FILES.md#cat): 
```md
String, including the ipfs handler, a cid and a path to traverse to, ie:
'/ipfs/QmXEmhrMpbVvTh61FNAxP9nU7ygVtyvZA8HZDUaqQCAb66'
'/ipfs/QmXEmhrMpbVvTh61FNAxP9nU7ygVtyvZA8HZDUaqQCAb66/a.txt'
'QmXEmhrMpbVvTh61FNAxP9nU7ygVtyvZA8HZDUaqQCAb66/a.txt'
```

Is there any actual difference between these options? Like does adding /ipfs/ search the whole network and excluding it just searches your node? 

Is this specific to the `cat` command or for all things retrieval-ish, like `get` and `ls` as well?

lidel replied:
Both work the same (we default to `/ipfs/` if missing), but I agree the tutorial should be explicit to remove any ambiguity:

Without `/ipfs/`  it is not obvious what `QmXEmhrMpbVvTh61FNAxP9nU7ygVtyvZA8HZDUaqQCAb66` is. 
It could be a CID under `/ipfs/`, but it ALSO be a libp2p-key under `/ipns/`.

terichadbourne replied:
Just adding this note based on a convo with @lidel: We should use `/ipfs/<CID>/<path>` whenever we're using a method that accepts CID/path, so `cat`, `get`, `ls`,  etc. It's only okay to use the CID alone if there is no path after it. In addition to disambiguating from libp2p as mentioned above it will make things more consistent if we later teach `/ipns/` paths. 

I'll accept the suggestion here and then look at the lessons again to see where else this pattern might need to be updated or explained.

dominguesgm commented on src/tutorials/0005/03.vue:
```diff
@@ -0,0 +1,109 @@
+      }
+    }
+
+    const lsResult = await pTimeout(ipfs.ls(resultData.hash), 2000).catch(() => 'error')
```
Awesome! When the version that introduces timeouts is released, I think that would be a relevant topic for an additional lesson.

dominguesgm commented on src/tutorials/0005/03.vue:
```diff
@@ -0,0 +1,109 @@
+      }
+    }
+
+    const lsResult = await pTimeout(ipfs.ls(resultData.hash), 2000).catch(() => 'error')
```
Good catch, thanks!

dominguesgm commented on src/tutorials/0005/04-exercise.md:
```diff
@@ -0,0 +1,9 @@
+
+Using the `cat` function, retrieve the file's contents and return them as a string.
+
+**Hint:** Don't forget to convert the buffered contents of the text file to a string using `.toString('utf8')` before returning your result.
```
The point is to convert the content of the file we placed in the node so it's readable by the user in the lesson output. Otherwise, the lesson output will only be a Buffer, which is arguably clunkier. This is something we already do in the MFS tutorial, too.

terichadbourne replied:
fixed for lesson 7 but left here

dominguesgm commented on src/tutorials/0005/04.md:
````diff
@@ -0,0 +1,23 @@
+In the previous lesson, you saw that each file added to IPFS has its own unique `hash` derived from its content. This `hash`, also known as a [CID (Content Identifier)](https://proto.school/#/data-structures/04), can be used like an address to access the file. If you know a file's CID, you can use the `cat` function provided by the Regular IPFS Files API -- similar to the one you may have seen pr ...
+
+```javascript
+await ipfs.cat(ipfsPath, [options], [callback])
````
Got it 👍

dominguesgm commented on src/tutorials/0005/05.vue:
```diff
@@ -0,0 +1,123 @@
+    fileObjectsArray.push(fileObject)
+  })
+
+  // Alternatively, using the Array.map method:
```
@terichadbourne do you want to wheigh in here?

terichadbourne replied:
We have a lot of JavaScript beginners who don't know specific Array methods, and they find it very hard to understand the methods that smush everything into one line, so we've been offering two options in the solution code, one of them commented out. I'm okay with swapping which one is presented as the alternative, but I'd rather not remove the easier-to-parse version.

terichadbourne replied:
this one is still outstanding @dominguesgm

dominguesgm commented on src/tutorials/0005/06-exercise.md:
```diff
@@ -0,0 +1,3 @@
+Use `ls` to list the contents of the root directory you created in the previous exercise.
+
+**Hint:** The CID of the root directory ( `""` ) has been stored for you in the `rootCID` variable in the starter code below. You could have jotted this value down after the last exercise, but here we're looking through the array returned by the `add` method to find the element with the empty string as its `path` (the root directory) and saving its `hash` value, or CID.
```
> You could have jotted this value down after the last exercise, but here we're looking at the last element (the wrapping directory) in the array returned by the `add` method and saving its `hash` value, or CID.

Do you think this would be better?

terichadbourne replied:
I'm fine with changing the way we search for it, but I don't think that's quite enough context to show that the last entry is ALWAYS the wrapping directory which is key to presenting this as a reasonable way to approach it. 

But regardless, I don't think this addresses the problem we were trying to anticipate. If you added some files using `{ wrapWithDirectory: true }` and later wanted to find them, how would you know where they were? Our statement was meant to be more about "write this down right now because you'll never see it except in the results of `add`." Is that incorrect? 

If I do `add`, wrapping with directories, 5 times, it creates 5 different I'm-not-calling-them-root directories. If I need to go back and retrieve a specific one of them later, while I don't have access to the results of my original `add` method, how do I refer to it?

terichadbourne commented on src/static/courses.json:
```diff
@@ -1,4 +1,4 @@
-  "all": ["0004", "0001", "0002", "0003"],
-  "featured": ["0004", "0001", "0002", "0003"]
+  "all": ["0004", "0005", "0001", "0002", "0003"],
+  "featured": ["0001", "0004", "0005", "0002"]
```
Haha. @fsdiogo proposed 3-digit IDs and I pushed back. What, does it not seem to you like our content growth rate justifies this future-proofing? A gal can dream. 😂

terichadbourne commented on src/tutorials/0005/03.md:
```diff
@@ -0,0 +1,43 @@
+
+## Working with files in ProtoSchool
+Here in our ProtoSchool tutorials, we create a new IPFS node for you in the browser each time you hit the "Submit" button in a lesson. Whenever you see `ipfs.someMethod()` in our lessons, `ipfs` is a variable that refers to your IPFS instance, also known as a node. The actions that you take only affect your own IPFS node, not nodes belonging to your peers.
```
Hmmm, so I have been under the impression that we're creating a new node in the browser each time they hit submit on a lesson and not actually labelling any files that are on their computer, not pinning anything, and neither they nor we are keeping a node live, so they are in effect not putting anything accessible on the network. As in, don't worry too much about which files you upload for these exercises, they aren't ending up on the gateway and no one knows enough to find them. When we give them challenges to access specific files we've made for them, we're also creating them locally in the node we create rather than putting them on the gateway because it was taking impossibly to infinitely long to grab them otherwise. I assume all of this stuff disappears as soon as they close their browser tab? This is the "special"ness of what I thought we were doing; we're making something even less permanent than it usually is. 

Somewhere else there's a note to the effect of "when you're ready to do this thing for real you'll need to learn how to install IPFS and host your stuff"

Was some of this wording  more accurate for the MFS lesson we copied it from? Like making directories and such affected no one but you? 

We can add some disambiguation here about adding being just to your own node but other methods searching your node first, then other nodes if not found there. If so, we could either do it here or wait and do it as we present each method. Or we could mention pinning/sharing here and link to some resource you'd recommend about that.

terichadbourne replied:
@alanshaw does this edit address your concerns? I propose using this text to clarify what I was trying to express here (the one-ephemeral-node-per-lesson concept) and then later address lesson-by-lesson where stuff goes or doesn't go when you use a specific method.
> Here in our ProtoSchool tutorials, we create a new IPFS node for you in the browser each time you hit the "Submit" button in a lesson. Whenever you see `ipfs.someMethod()` in our lessons, `ipfs` is a variable that refers to your IPFS instance, also known as a node. This IPFS node is not carried over from one lesson to another, which is why you'll see that we often prepopulate some code in an exercise to get your new IPFS node to the state you got a different IPFS node to in a previous one.
> 
> We create these IPFS nodes behind the scenes so that you can focus on the content of each lesson.  Outside of ProtoSchool, though, you'd want a more consistent experience and a node (or multiple nodes) that you could access repeatedly. To acheive this, you could either initialize [JS IPFS in the browser](https://github.com/ipfs/js-ipfs#use-in-the-browser) on your own, or host your own node locally by installing IPFS and running a daemon in your terminal. When you're ready to experiment with that, you can visit the IPFS docs site to learn how to [install IPFS](https://docs.ipfs.io/guides/guides/install/) and [initialize your node](https://docs.ipfs.io/introduction/usage/#initialize-the-repository).

terichadbourne replied:
Added this ^ plus individual updates to method lesson in my latest commits. Marking resolved but we can address if you think any of my edits don't do the concepts justice.

terichadbourne commented on src/tutorials/0005/04.md:
```diff
@@ -0,0 +1,23 @@
+
+An IPFS `path` can take several formats (you can read about them in the [Files API](https://github.com/ipfs/interface-js-ipfs-core/blob/master/SPEC/FILES.md#cat) documentation). In this lesson, we'll use the simplest one: the `hash` string (CID) present in the object we receive for a file when the `ipfs.add` function is called. You may remember from the previous lesson that the `add` method retur ...
+
+Data returned by the `cat` method comes in the form of a `Buffer`. If the file being retrieved is a plain text file such as a `.txt`, you can convert its buffered contents to a string by calling the JavaScript method `.toString('utf8')`.
```
I think it's a great idea to add that.

terichadbourne replied:
done in latest commit

terichadbourne commented on src/tutorials/0005/05.md:
```diff
@@ -0,0 +1,69 @@
+
+This is because of the way in which IPFS uses [content addressing](https://proto.school/#/data-structures/03): different contents lead to a different cryptograhpic hash (Content Identifier), whether we're working with directories or files themselves.
+
+But what if you forgot to add a file to a directory you just created? You'll have to call the `add` method again with all the files you want wrapped with that directory, resulting in a new directory altogether with a new CID.
```
Love the idea of adding this point.

terichadbourne replied:
Is this reasonably accurate @alanshaw? 

> But what if you forgot to add a file to a directory you just created? You'll have to call the `add` method again with all the files you want wrapped with that directory, resulting in a new directory altogether with a new CID.
> 
> However, it's important to note that the file won't actually be stored multiple times if you do this. Unlike on your computer, where you can accidentally download the same file twice, store it in two different directories, and double your storage needs, the IPFS network inherently knows the contents of a file based on its CID, and can therefore keep only a single copy of the data that the CID refers to. It's only the reference to that data (the CID, which you can think of as a link or an addresses), which is stored in each of the directories you create. This efficiency is one of IPFS's key advantages.
> 
> **Rather than thinking of directories created with `{ wrapWithDirectory: true }` as traditional file folders, it's more useful to think of them as naming shortcuts, as we'll see in upcoming lessons.**

terichadbourne replied:
that's been added ^ please let us know if you think it needs edits

terichadbourne commented on src/tutorials/0005/07-exercise.md:
```diff
@@ -0,0 +1,17 @@
+
+There are two valid ways to solve this challenge. Take your pick!
+
+**Hint:** Remember to convert the `Buffer` returned by `cat` to a string, as you did previously, using `.toString('utf-8')`.
```
I'm updating the code in lesson 7 to return the stringified value to save them the step.

terichadbourne commented on src/tutorials/0005/07-exercise.md:
```diff
@@ -0,0 +1,17 @@
+
+Based on our file structure, the path could be expressed as:
+
+* `dir/success.txt`, relative to the root directory
```
I support this idea but it will require some changes under the hood to make a new file with a new message hidden it that's used in lessons 7 and 8 and doesn't take up too much space when the buffer is shown on the screen in lesson 8. 

Are fun emojis super space consuming? Or  how about including this kind of emoji-ish thing in a more fun version?--->   ¯\_(ツ)_/¯ 

 I don't have time to get to it today so I'll leave this to @dominguesgm for early Monday.

terichadbourne commented:
@dominguesgm The collapsing conversations make it kind of hard to keep track of, but I'm pretty sure I've gotten to all the suggestions you hadn't gotten to yet except for: 
- swapping map versus whatever else we had as the recommended solution, leaving the beginner-friendly one visible as an alternative solution
- making a friendlier text file

I also went ahead and merged my PR and then pulled it into this PR to get API to be properly capitalized in the shortname. 

@alanshaw @lidel if you have time Monday morning to take a quick look at today's commits, which include me addressing a number of issues you suggested from a text standpoint, that would be awesome.

lidel commented on src/tutorials/0005/07-exercise.md:
```diff
@@ -0,0 +1,19 @@
+
+**Hint:** In the code below we've already taken care of converting the `bufferedContents` to a string before returning the result, as you did yourself previously using `.toString()`.
+
+Don't forget to preapend `/ipfs/` to the IPFS Path string
```
```suggestion
Don't forget to prepend `/ipfs/` to the IPFS Path string
```

dominguesgm replied:
I think it doesn't, because when finding the `{`, it expects a block in this situation and not an object literal. It throws an `Error: unexpected token: ':'`

This would work, however:

```javascript
const fileObjectsArray = files.map((file) => ({ path: file.name, content: file }))
```

But I'm not sure it is an improvement in terms of readability.

lidel commented on src/tutorials/0005/06.vue:
```diff
@@ -0,0 +1,113 @@
+
+const code = `/* global ipfs */
+const run = async (files) => {
+  const fileObjectsArray = files.map((file) => { return { path: file.name, content: file }})
```
I think shorter notation should work? (here and in line 32)
```suggestion
  const fileObjectsArray = files.map((file) => { path: file.name, content: file })
```

terichadbourne commented:
@dominguesgm I've made some tweaks to validation messages. Please see the comments in lessons 5 and 7 in the places where I think the validation or messages associated with it still needs some tweaking. I also replaced the text of your catch-all message with something more action-oriented. This feedback is mostly from me reading your validation code and not from going through and trying to break it by doing things wrong that you didn't anticipate   (which I'm generally great at but haven't had time for yet). 

I also went through and changed all references to "root directory" to "top-level directory" per some earlier feedback from @alanshaw, and I moved your links to array methods to the hints section so people can see them before failing rather than after giving up and peeking at the solution. :)

@hacdias if you happen to have any time tomorrow morning to take a quick look at the validation code or try submitting some various predictably-wrong code and seeing if it catches on meaningful errors, that would be awesome.  

Hoping to get this published by end of US day tomorrow if no one has lingering concerns.
